procfile/procfile.go

```diff
@@ -0,0 +1,138 @@
+// Package procfile contains methods for parsing standard and extended
+// Procfiles.
+package procfile
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+
+	"gopkg.in/yaml.v2"
+)
+
+// ErrBadProcfile is returned by Parse when the given procfile is invalid.
+var ErrBadProcfile = errors.New("procfile: unknown Procfile format")
+
+// StandardProcfile represents the standard Procfile format, without extended
+// atrributes.
+type StandardProcfile map[string]string
+
+func (p StandardProcfile) ProcessDefinitions() (definitions []ProcessDefinition, err error) {
+	for k, command := range p {
+		definitions = append(definitions, ProcessDefinition{
+			Name:    k,
+			Command: command,
+		})
+	}
+	return
+}
+
+// ExtendedProcfile represents the extended Procfile format. This format is also
+// compatible with the docker-compose.yml format.
+type ExtendedProcfile map[string]struct {
+	Command      string                   `yaml:"command"`
+	HealthChecks []map[string]interface{} `yaml:"health_checks"`
+}
+
+func (p ExtendedProcfile) ProcessDefinitions() (definitions []ProcessDefinition, err error) {
+	for k, pd := range p {
+		var healthChecks []HealthCheck
+		for _, v := range pd.HealthChecks {
+			t := v["type"]
+			switch t {
+			case "http":
+				var h HealthCheck
+				h, err = newHTTPHealthCheck(v)
+				if err != nil {
+					return
+				}
+				healthChecks = append(healthChecks, h)
+			case "tcp":
+				healthChecks = append(healthChecks, TCPHealthCheck{})
+			default:
+				err = fmt.Errorf("unknown health check type: %s", t)
+				return
+			}
+		}
+
+		definitions = append(definitions, ProcessDefinition{
+			Name:         k,
+			Command:      pd.Command,
+			HealthChecks: healthChecks,
+		})
+	}
+	return
+}
+
+// ProcessDefinition defines parameters about individual processes within the
+// Procfile.
+type ProcessDefinition struct {
```

Comment:
I was also thinking of defining a `twelvefactor` package, that defines various types like:

```go
// We define an app as a list of processes that share an environment.
type App struct {
        Env map[string]string
        Processes []Process
}

type Process struct {
        Name string
        Command string
        // etc
}
```

then scheduler.Scheduler is a means of running `twelvefactor.App`, which is pretty cool.